jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -159,10 +318,332 @@ default void clear()
         BufferUtil.clear(getByteBuffer());
     }
 
+    /**
+     * <p>Skips, advancing the ByteBuffer position, the given number of bytes.</p>
+     *
+     * @param length the maximum number of bytes to skip
+     * @return the number of bytes actually skipped
+     */
+    default long skip(long length)
+    {
+        if (length == 0)
+            return 0;
+        ByteBuffer byteBuffer = getByteBuffer();
+        length = Math.min(byteBuffer.remaining(), length);
+        byteBuffer.position(byteBuffer.position() + Math.toIntExact(length));
+        return length;
+    }
+
+    /**
+     * <p>Limit this buffer's contents to the size.</p>
+     *
+     * @param size the new size of the buffer
+     */
+    default void limit(long size)
+    {
+        ByteBuffer byteBuffer = getByteBuffer();
+        size = Math.min(size, byteBuffer.remaining());
+        byteBuffer.limit(byteBuffer.position() + Math.toIntExact(size));
+    }
+
+    /**
+     * Get a slice of the buffer.
+     * @return A sliced {@link RetainableByteBuffer} sharing this buffers data and reference count, but
+     *         with independent position. The buffer is {@link #retain() retained} by this call.
+     * @see #slice(long)
+     */
+    default RetainableByteBuffer slice()
+    {
+        return slice(Long.MAX_VALUE);
+    }
+
+    /**
+     * Get a partial slice of the buffer.
+     * This is equivalent to {@link #slice()}.{@link #limit(long)}, but may be implemented more efficiently.
+     * @param length The number of bytes to slice, which may beyond the limit and less than the capacity, in which case
+     * it will ensure some spare capacity in the slice.
+     * @return A sliced {@link RetainableByteBuffer} sharing the first {@code length} bytes of this buffers data and
+     * reference count, but with independent position. The buffer is {@link #retain() retained} by this call.
+     */
+    default RetainableByteBuffer slice(long length)
+    {
+        int size = remaining();
+        ByteBuffer byteBuffer = getByteBuffer();
+        int limit = byteBuffer.limit();
+
+        byteBuffer.limit(byteBuffer.position() + Math.toIntExact(Math.min(length, size)));
+        ByteBuffer slice = byteBuffer.slice();
+        byteBuffer.limit(limit);
+        if (length > size)
+            slice.limit(size);
+
+        if (!canRetain())
+            return new NonRetainableByteBuffer(slice);
+
+        retain();
```

Comment:
It's good that you javadoc'ed that the originating RBB gets retained by slice(), but I remember having a lengthy discussion about slice() retaining or not with @sbordet, but I cannot remember the outcome, so we may need to revisit this later.

Reply:
This approach is working for this PR, so will keep unless you rethink what the issue was.